CE-915 - Consul TLS with vault (#1359)
hashicorp/web-unified-docs: content/consul/v1.22.x/content/docs/automate/consul-template/vault/mtls.mdx

- **Vault:** you need a running Vault cluster in your network. You can use a [local Vault dev server](/vault/tutorials/get-started/setup#set-up-the-lab) or an existing Vault deployment. You also need to configure your terminal to interact with Vault cluster by setting `VAULT_ADDR`, `VAULT_CACERT`, and `VAUKT_TOKEN`.

- **Consul Template:** to automate certificate generation and distribution you need [the `consul-template` binary installed](/consul/docs/automate/consul-template/install) on the agent's node.

The diagram below shows the minimal architecture needed to demonstrate the functionality.

![Architectural diagram showing a Client server and a Vault server with an operator issuing a command to start an automation](/img/consul-template/consul-vault-tls.png)

To configure Vault as Consul's certification authority you need to follow these steps:

1. [Configure the root certification authority (CA).](#configure-the-root-certification-authority-ca)
1. [Configure an intermediate CA.](#configure-an-intermediate-ca)
1. [Create a Vault role.](#create-a-vault-role)
1. [Generate certificates for Consul.](#generate-certificates-for-consul)

> 1. (Optional) [Automate certificate rotation with consul-template.](#automate-certificate-rotation-with-consul-template)

Reviewer: ```suggestion
1. [Configure the root certification authority (CA)](#configure-the-root-certification-authority-ca)
1. [Configure an intermediate CA](#configure-an-intermediate-ca)
1. [Create a Vault role](#create-a-vault-role)
1. [Generate certificates for Consul](#generate-certificates-for-consul)
1. (Optional) [Automate certificate rotation with consul-template](#automate-certificate-rotation-with-consul-template)
```

# Generate mTLS Certificates for Consul with Vault and Consul Template

This page describes how to use Vault's [PKI Secrets Engine](/vault/docs/secrets/pki) to generate and renew dynamic X.509 certificates, using [Consul Template](/consul/docs/automate/consul-template) to automatically rotate your certificates.

This method enables each agent in the Consul datacenter to have a unique certificate, with a relatively short time-to-live (TTL), that is automatically rotated, which allows you to safely and securely scale your datacenter while using mutual TLS (mTLS).

If you want to use Vault as Consul service mesh certification authority, refer to [Vault as Consul service mesh certification authority](/consul/tutorials/operate-consul/vault-pki-consul-connect-ca).

## Prerequisites

- **Consul:** you need at least one Consul server node and, ideally, a Consul client agent node. Follow [Deploy Consul on VMs](/consul/tutorials/get-started-vms/virtual-machine-gs-deploy) to learn how to deploy a Consul agent. This page will provide you with the necessary specific configuration to apply for the scenario.

> - **Vault:** you need a running Vault cluster in your network. You can use a [local Vault dev server](/vault/tutorials/get-started/setup#set-up-the-lab) or an existing Vault deployment. You also need to configure your terminal to interact with Vault cluster by setting `VAULT_ADDR`, `VAULT_CACERT`, and `VAUKT_TOKEN`.

Author: ```suggestion
- **Vault:** you need a running Vault cluster in your network. You can use a [local Vault dev server](/vault/tutorials/get-started/setup#set-up-the-lab) or an existing Vault deployment. You also need to configure your terminal to interact with Vault cluster by setting `VAULT_ADDR`, `VAULT_CACERT`, and `VAULT_TOKEN`.
```

Verify the rotation by checking that Consul Template keeps listing, every 24 hours, a timestamp and the log line:

```plaintext hideClipboard
Configuration reload triggered
```

Use `openssl` to verify the certificate content:

```shell-session
$ openssl x509 -text -noout -in /etc/consul.d/agent.crt
```

Verify in the output that the `Not Before` and `Not After` values are being updated to reflect the new certificate.

<CodeBlockConfig hideClipboard highlight="9-10">

```plaintext
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            0a:06:b2:62:56:d1:7d:e6:5c:67:f7:89:06:8a:3e:5b:f1:fa:7e:00
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN=dc1.consul Intermediate Authority
        Validity
            Not Before: Nov 25 11:48:06 2025 GMT
            Not After : Nov 25 12:48:06 2025 GMT
        Subject: CN=server.dc1.consul
## ...
```

</CodeBlockConfig>

## Next steps

In this page, you learned how to generate Consul TLS certificates using Vault's PKI secret engine and to automatically rotate them using consul-template.

> If you want to automate Consul's gossip encryption using Vault and consul-template refer to [Generate and manage gossip encryption for Consul with Vault and Consul Template](/consul/docs/automate/consul-template/vault/gossip).

Reviewer: ```suggestion
If you want to automate Consul's gossip encryption using Vault and `consul-template` refer to [Generate and manage gossip encryption for Consul with Vault and Consul Template](/consul/docs/automate/consul-template/vault/gossip).
```

> To learn how to use Vault to manage Consul's ACL token generation refer to [Use Vault for ACL management with Consul on VMs](/consul/docs/secure/acl/vault/vm).

Reviewer: ```suggestion
To learn how to use Vault to generate and manage Consul's ACL tokens, refer to [Use Vault for ACL management with Consul on VMs](/consul/docs/secure/acl/vault/vm).
```
